### PR TITLE
Add Publish Preview dialog

### PR DESCRIPTION
**Pull request by infinite-persistence**
_Requires a change in lbry-redux_ ([#344 doPublish: Add optional parameter for "preview-only" request.](https://github.com/lbryio/lbry-redux/pull/344))

## Feature
A no-scroll, single-glance dialog to preview the Publish data before really uploading.
Includes estimated transaction fees.

image

## Issue
Closes [#486 Add Confirmation before publishing file or creating channel (show fee estimate)](https://github.com/lbryio/lbry-desktop/issues/486)

## Changes
- See commit messages for walkthrough.

## Testing notes
Probably not useful, but hopefully can spark any ideas on missing test areas.
<details>
  <summary>Click to expand</summary>

Issues:
[x] Trancoding started even in preview; it should not. ==> force to "false" in preview-mode.
[x]    Test: trancoding still works after disabling in preview (ensure not permanently disabled)
[x] Tags that reflowed to second line need gaps.
[x] Long tags need to be truncated with ellipsis.
[x] Long tag touches border.
[ ] Warning: "each child in list should have unique key". ==> No longer reproducible.
[x] Warning: "whitespace text nodes cannot appears as child nodes in <tbody>" ==> Firefox didn't like the empty row at the bottom. Just removed it. We'll just do without the final separator line (which is probably what the theme intended anyways).

Todo:
[x] Include tags
[x] Include Transcode section
[x] Test web
[x]    The preview is not appearing. ==> `filePath` in web is a `WebFile` object. Need to handle it differently. Not sure how to get the full path in web, so we'll just show the filename for now.
[x]    The preview modal takes a while to appear in web. ==> Grayed out the button while previewing.
[x] Handle "license=other"
[x] Handle "license=copyrighted"
[x] Test: empty description
[x] Test on mobile screen
[x]    Issue: the table is overflowing on the right. ==> Using `nowrap` on the first column causes this. Just fix the label to 30% and truncate with ellipsis.
[x] Extreme case: What if screen height is super small? ==> it scrolls, so no problem.  Defeats the no-scroll intention, but at least it's not in an unusable state.
[x] Description is markdown. How to show? ==> just use `<MarkdownPreview>`
[x] What if Description is too long? ==> made row scrollable if exceeded 5-6 lines.
[x] What if `previewResponse` is null? ==> handled the null case, although not sure if it will ever happen.
[x] Test: Unplug cable ==> Preview doesn't seem to need internet connection, so nothing to handle.
[x] Test: Didn't break actual publish ==> Seems ok, as far as I can tell.  The same data is still sent to doPublish per original code.
[x] What if Editing instead of Publishing? ==> Disable preview for edits.
</details>

### CODE REVIEW

**Comment by seanyesmunt** [comment]
This is great! I will probably add a few commits to this to change the styling a bit since we have already planned what this should look like somewhat, but that will be much easier now that all of the redux code is set up. I probably won't get to it until later this week though.

**Comment by infinite-persistence** [comment]
Ah, it's part of a planned work.  Feel free to use as much or as little of it; I don't mind it being dropped entirely if it's not aligned to planned work 👍

**Comment by seanyesmunt** [comment]
By planned I mean loosely discussed a couple of times. You can take a shot at it if you want but the main thing I would be interested in changing is making this less "technical" and more a simplified version of what someone would see on the file page.

**Comment by infinite-persistence** [comment]
Rebasing to fix conflict with "in-app markdown post":

**Comment by tzarebczan** [comment]
Should we allow users to skip confirmation if they don't wish to see it each time? Super excited for this, especially around the estimated fees!

**Comment by infinite-persistence** [comment]
`Change:`  Added setting to never show the preview modal.  It can be re-enabled in the Advanced Settings.
The text and position are stubs since the UI is subject to change.
<details>
  <summary><b>Screenshots</b> (click to expand)</summary>

image
image
</details>

File: ui/scss/component/_table.scss [hunk 1]
```diff
@@ -204,3 +204,29 @@ th {
     margin-left: var(--spacing-m);
   }
 }
+
+.table--publish-preview {
```
Comment (kauffj) [inline ui/scss/component/_table.scss:208]: this should not be a new style

table--file-details and table--publish-preview should probably be unified @seanyesmunt

File: ui/modal/modalPublishPreview/view.jsx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+              <>
+                <div className="section__actions">
+                  <Button autoFocus button="primary" label={__('Publish')} onClick={() => this.onConfirmed()} />
+                  <Button button="link" label={__('Cancel')} onClick={closeModal} />
+                </div>
+                <p className="help">{__('Once the transaction is sent, it cannot be reversed.')}</p>
+                <FormField
+                  type="checkbox"
+                  name="sync_toggle"
+                  label={__("Don't show previews anymore.")}
```
Comment (kauffj) [inline ui/modal/modalPublishPreview/view.jsx:165]: needs better text

maybe "Publish without this preview", but I don't love it

File: ui/component/publishSettings/view.jsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  function handleChange() {
+    setEnablePublishPreview(!enablePublishPreview);
+  }
+
+  return (
+    <div>
+      <FormField
+        type="checkbox"
+        name="sync_toggle"
+        label={__('Review changes before uploading.')}
```
Comment (kauffj) [inline ui/component/publishSettings/view.jsx:23]: this text should match or be similar to the text that allows toggling it at the bottom of the confirmation dialog

File: ui/modal/modalPublishPreview/view.jsx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  createRow(label: string, value: any) {
+    // Probably need to do the colon differently when in RTL mode. Don't want it to be part of the string
+    // to reduce redundant strings for translation.
+    return (
+      <tr>
+        <td>
+          {label}
+          {label ? ':' : ''}
```
Comment (kauffj) [inline ui/modal/modalPublishPreview/view.jsx:61]: colons are an unnecessary here, and generally should be avoided outside of visually indistinct text

**Comment by infinite-persistence** [comment]
`Change`
- Use "Skip preview and confirmation" in both Modal and Settings for better correlation.
   - _Text is still subject to change until a better one comes along._
- Remove unnecessary colon in the Label column
- Enable "Publish Preview" on Edit Mode as well.
- Enable "Publish Preview" for Markdown.
   - _Depending on what the Preview Modal will be showing, we might need to pass the 'mode' over so that we can hide items like 'Description' when doing Markdown. As I'm not sure if the Description will be part of the modal or not, I skipped this for now._

**Comment by infinite-persistence** [comment]
`Rebase`
Rebased to grab the lint fixes for _publishForm.jsx_

**Comment by kauffj** [comment]
@infinite-persistence is there more work required here? Are you coming back to this?

**Comment by infinite-persistence** [comment]
> @infinite-persistence is there more work required here? Are you coming back to this?

I think 2 things to do:
1. Rebase to fix any conflicts with the new MD Editor (probably not much)
2. GUI changes that Sean had in mind.

I mulled over item-2 for a while, but not sure how I should change it.  I kinda like the existing "single glance to see all entered info" version to double-check before posting.

**Comment by seanyesmunt** [comment]
I'm fine merging this with the current styling, but disabling on Odysee. Lets wait until we release the next version of the app first though.

**Comment by infinite-persistence** [comment]
`Final changes`
- Rebased to latest master.
- Simplified the preview by just showing important items:  _title_, _URL_, _deposit_, _price_, _transcode_ and _estimated_cost_.

`Issue`
- On Web, the preview takes forever to appear if the file is large. I believe the file is uploaded by `apiPublishCallViaWeb`.  I'm not sure if the file is needed to calculate the cost or not, nor was I able to figure out how to prevent the full upload.  
I don't think this is usable if we can't calculate the cost locally on the browser (the upload time is doubled).

Considering abandoning this...

**Comment by seanyesmunt** [comment]
@infinite-persistence could we just not calculate the cost? I think it's still helpful to confirm all of the publish values.

**Comment by infinite-persistence** [comment]
> _@infinite-persistence could we just not calculate the cost?_

Ah, given #486 [Add Confirmation before publishing file or creating channel (show fee estimate)](https://github.com/lbryio/lbry-desktop/issues/486), I thought that's the primary reason to add the dialog.
Ok, I'll remove the cost field;   solves the double-upload problem on Web.

**Comment by seanyesmunt** [comment]
It would be nice to have, but if it's super slow I don't think it's necessary.

**Comment by infinite-persistence** [comment]
Would "do it on Desktop only" be a good idea?

**Comment by infinite-persistence** [comment]
`Changes`
- Disable 'tx-fee' field to handle the web delay problem.
- Undo the simplification -- all fields will be previewed.
- Rebranding update: "Publish" --> "Upload";  Use LBC icon.
- Hide irrelevant fields (`File`) in MD-POST mode.

**Comment by tzarebczan** [comment]
I wonder if we can just send a temp file up (like we do with post publishing) on web for the preview, and then send the actual file during confirmation. The file itself is still needed, but it doesn't matter what file is sent as this does not affect the cost. The main factor for tx cost is the length of the URL

**Comment by infinite-persistence** [comment]
> _I wonder if we can just send a temp file up (like we do with post publishing) on web for the preview, and then send the actual file during confirmation. The file itself is still needed, but it doesn't matter what file is sent as this does not affect the cost. The main factor for tx cost is the length of the URL_

It didn't work for me yesterday, but after seeing your comment, I gave it another shot and discover my syntax mistake.  It works!
It still takes a few seconds compared to Desktop, but not a very long wait.

Fix coming up...

**Comment by infinite-persistence** [comment]
`Change`
- Send a dummy file on Web preview to fix the delay problem.

**Comment by seanyesmunt** [comment]
Awesome!

**Pull request merged**